docs/source/developer/threading_implementation.rst

```diff
@@ -0,0 +1,197 @@
+=========================================
+Notes on Numba's threading implementation
+=========================================
+
+The execution of the work presented by the Numba ``parallel`` targets is
+undertaken by the Numba threading layer. Practically, the "threading layer"
+is a Numba built-in library that can perform the required concurrent execution.
+At the time of writing there are three threading layers available, each
+implemented via a different lower level native threading library. More
+information on the threading layers and appropriate selection of a threading
+layer for a given application/system can be found in the
+:ref:`threading layer documentation <numba-threading-layer>`.
+
+The pertinent information to note for the following sections is that the
+function in the threading library that performs the parallel execution is the
+``parallel_for`` function. The job of this function is to both orchestrate and
+execute the parallel tasks.
+
+Thread masking
+--------------
+
+In order to simplify the design, it was decided that Numba should never launch
+new threads beyond the threads that are launched initially with
+``_launch_threads`` when the first parallel execution is run. Consequently,
+the programmatic setting of the number of threads can only be done by setting
+the number of threads to a number less than the total number that have already
+been launched. This is done by "masking" out unused threads, causing them to
+do no work. For example, on a 16 core machine, if the user were to call
+``set_num_threads(4)``, numba would always have 16 threads present, but 12 of
+them would sit idle for parallel computations. A further call to
+``set_num_threads(16)`` would cause those same threads to do work in later
+computations.
+
+:ref:`Thread masking <numba-threading-layer-thread-masking>` was added to make
+it possible for a user to programmatically alter the number of threads
+performing work in the threading layer. Thread masking proved challenging to
+implement as it required the development of a programming model that is suitable
+for users, easy to reason about, and could be implemented safely, with
+consistent behavior across the various threading layers.
+
+Programming model
+~~~~~~~~~~~~~~~~~
+
+The programming model chosen is similar to that found in OpenMP. The reasons
+for this choice were that it is familiar to a lot of users, restricted in
+scope and also simple. The number of threads in use is specified by calling
+``set_num_threads`` and the number of threads in use can be queried by calling
+``get_num_threads``.These two functions are synonymous with their OpenMP
+counterparts (with the above restriction that the mask must be <= the number
+of launched threads). The execution semantics are also similar to OpenmP in
+that once a parallel region is launched altering the thread mask, it has no
+impact on the currently executing region but will have an impact on parallel
+regions executed subsequently.
```

Comment:
I'm not sure that this makes sense. The launch of a parallel region has no impact on the thread mask, the original text was correct:
```
The execution semantic is also similar to OpenMP in that once a
parallel region is launched, altering the thread mask has no impact on the
currently executing region, but will have an impact on parallel regions executed
subsequently.
```
?

i.e. once in a `parallel for` loop, setting the thread mask won't do anything to the executing loop, but a subsequent (logical or nested) `parallel for` loop will adhere to the new mask.

Reply:
Oh my bad. I think the missing comma before "altering" confused me.

Reply:
Ah, probably my bad, I should have put the comma in in the first place! Anyway, looks good now, thanks. Sorry for the confusion!